grpc/shared_conn.go

```diff
@@ -18,68 +24,190 @@ type OnTrackCB func(tr *webrtc.TrackRemote, r *webrtc.RTPReceiver)
 // For modules, the grpc connection is over a Unix socket. The WebRTC `PeerConnection` is made
 // separately. The `SharedConn` continues to implement the `rpc.ClientConn` interface by pairing up
 // the two underlying connections a client may want to communicate over.
+//
+// The lifetime of SharedConn objects are a bit nuanced. SharedConn objects are expected to be
+// handed to resource Client objects. When connections break and become reestablished, those client
+// objects are not recreated, but rather we swap in new connection objects under the hood.
+//
+// This is because resource management controls updating the graph nodes with Client objects only
+// when calls to `Reconfigure` are made. But connections can break and heal without any calls to
+// Reconfigure.
+//
+// Thus the lifetime of a SharedConn for a client is _not_:
+// 1) Object initialize
+// 2) Working
+// 3) Close
+// 4) Object destruction
+// And back to initialization.
+//
+// But rather:
+// 1) Object initialization
+// 2) `ReplaceConn`
+// 3) `Close`
+// 4) `ReplaceConn`
+// ...
+// 5) Object destruction (at shutdown or resource removed from config).
 type SharedConn struct {
-	*ReconfigurableClientConn
-	peerConn           *webrtc.PeerConnection
-	logger             logging.Logger
-	mu                 sync.RWMutex
-	resourceOnTrackCBs map[resource.Name]OnTrackCB
+	grpcConn      ReconfigurableClientConn
+	peerConn      *webrtc.PeerConnection
+	PeerConnReady <-chan struct{}
```

Comment:
It looks like the only user of `PeerConnReady` are methods of `SharedConn`. Why is it public?

Reply:
You're right. IIRC, this was moved from "wait when adding a resource" at the last minute. Just to simplify the question of when the waiting happens. But I didn't update the publicness to reflect that.